wordapi_rework/script_extraction.py

```diff
@@ -0,0 +1,187 @@
+            lvlText = lvl.find('w:lvlText', NS).attrib.get(f'{{{NS["w"]}}}val')
+            start = int(lvl.find('w:start', NS).attrib.get(f'{{{NS["w"]}}}val', '1')) if lvl.find('w:start', NS) is not None else 1
+            levels[ilvl] = {'numFmt': numFmt, 'lvlText': lvlText, 'start': start}
+        abstractNumId_to_levels[abstractId] = levels
+
+    return numId_to_abstractNumId, abstractNumId_to_levels
+
+def format_number(numFmt, value):
+    if numFmt == 'decimal':
+        return str(value)
+    elif numFmt == 'lowerLetter':
+        return chr(96 + value)
+    elif numFmt == 'upperLetter':
+        return chr(64 + value)
+    elif numFmt == 'lowerRoman':
+        return to_roman(value).lower()
+    elif numFmt == 'upperRoman':
+        return to_roman(value)
+    elif numFmt == 'bullet':
+        return '•'
+    else:
+        return str(value)
+
+def to_roman(n):
+    val = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
+    syms = ['M', 'CM', 'D', 'CD', 'C', 'XC', 'L', 'XL', 'X', 'IX', 'V', 'IV', 'I']
+    roman = ''
+    i = 0
+    while n > 0:
+        for _ in range(n // val[i]):
+            roman += syms[i]
+            n -= val[i]
+        i += 1
+    return roman
+
+def get_style_num_map(styles_root, numId_to_abstract, abstract_to_levels, num_root):
+    style_num_map = {}
+    for style in styles_root.findall('w:style', NS):
+        if style.attrib.get(f'{{{NS["w"]}}}type') != 'paragraph':
+            continue
+        style_id = style.attrib.get(f'{{{NS["w"]}}}styleId')
+        pPr = style.find('w:pPr', NS)
+        if pPr is not None:
+            numPr = pPr.find('w:numPr', NS)
+            if numPr is not None:
+                ilvl_el = numPr.find('w:ilvl', NS)
+                numId_el = numPr.find('w:numId', NS)
+                if ilvl_el is not None and numId_el is not None:
+                    ilvl = int(ilvl_el.attrib.get(f'{{{NS["w"]}}}val'))
+                    numId = int(numId_el.attrib.get(f'{{{NS["w"]}}}val'))
+                    style_num_map[style_id] = {'numId': numId, 'ilvl': ilvl}
+
+    for numId, abstractId in numId_to_abstract.items():
+        levels = abstract_to_levels.get(abstractId, {})
+        for ilvl in levels:
+            for absnum in num_root.findall('w:abstractNum', NS):
+                if absnum.attrib.get(f'{{{NS["w"]}}}abstractNumId') == str(abstractId):
+                    lvl_elem = absnum.find(f'./w:lvl[@w:ilvl="{ilvl}"]', NS)
+                    if lvl_elem is not None:
+                        pStyle_el = lvl_elem.find('w:pStyle', NS)
+                        if pStyle_el is not None:
+                            style_id = pStyle_el.attrib.get(f'{{{NS["w"]}}}val')
+                            style_num_map[style_id] = {'numId': numId, 'ilvl': ilvl}
+    return style_num_map
+
+def extract_paragraph_text(para):
+    return ''.join([node.text or '' for node in para.findall('.//w:t', NS)]).strip()
```

Comment:
Also, you can remove the `[ ... ]` to make it a generator comprehension and avoid creating a list in memory.
```suggestion
    return ''.join(node.text or '' for node in para.findall('.//w:t', NS)).strip()
```